src/value/builders.rs

```diff
@@ -0,0 +1,453 @@
+            .as_list()
+            .unwrap()
+            .clone_builder()
+            .remove(1)
+            .push(88)
+            .build();
+        let expected_list = ion!(r#"[1, "foo", bar, 88]"#);
+        assert_eq!(new_list, expected_list);
+    }
+
+    #[test]
+    fn build_sexp() {
+        let actual: Element = SExpBuilder::new()
+            .push(1)
+            .push(true)
+            .push("foo")
+            .push(Symbol::owned("bar"))
+            .build();
+        let expected = ion!(r#"(1 true "foo" bar)"#);
+        assert_eq!(actual, expected);
+    }
+
+    #[test]
+    fn sexp_clone_builder() {
+        let original_sexp = ion_sexp!(1 true "foo" Symbol::owned("bar"));
+        let new_sexp = original_sexp
+            .as_sexp()
+            .unwrap()
+            .clone_builder()
+            .remove(1)
+            .push(88)
+            .build();
+        let expected_sexp = ion!(r#"(1 "foo" bar 88)"#);
+        assert_eq!(new_sexp, expected_sexp);
+    }
+
+    #[test]
+    fn build_sexp_remove() {
+        let actual: Element = SExpBuilder::new()
+            .push(1)
+            .push(true)
+            .push("foo")
+            .remove(1)
+            .remove(1)
+            .push(Symbol::owned("bar"))
+            .build();
+        let expected = ion!("(1 bar)");
+        assert_eq!(actual, expected);
+    }
+
+    #[test]
+    fn build_sexp_with_macro() {
+        let actual: Element = ion_sexp!(1 true "foo" Symbol::owned("bar"));
+        let expected = ion!(r#"(1 true "foo" bar)"#);
+        assert_eq!(actual, expected);
+    }
+
+    #[test]
+    fn build_struct() {
+        let actual: Element = StructBuilder::new()
+            .with_field("a", 1)
+            .with_field("b", true)
+            .with_field("c", "foo")
+            .with_field("d", Symbol::owned("bar"))
+            .build();
+        let expected = ion!(r#"{a: 1, b: true, c: "foo", d: bar}"#);
+        assert_eq!(actual, expected);
+    }
+
+    #[test]
+    fn build_struct_with_macro() {
+        let actual: Element = ion_struct! {
+            "a": 1,
+            "b": true,
+            "c": "foo",
+            "d": Symbol::owned("bar")
+        };
+        let expected = ion!(r#"{a: 1, b: true, c: "foo", d: bar}"#);
+        assert_eq!(actual, expected);
+    }
+
+    #[test]
+    fn build_struct_without() {
+        let actual: Element = StructBuilder::new()
+            .with_field("a", 1)
+            .with_field("b", true)
+            .with_field("c", "foo")
+            .with_field("d", Symbol::owned("bar"))
+            .with_field("d", Symbol::owned("baz"))
+            // Removes the only 'b' field
+            .remove_field("b")
+            // Removes the first 'd' field, leaves the second
+            .remove_field("d")
+            .build();
+        let expected = ion!(r#"{a: 1, c: "foo", d: baz}"#);
+        assert_eq!(actual, expected);
+        assert_eq!(actual, expected);
+    }
+}
```

Comment:
🗺️ I can't comment directly on `element_stream_reader.rs` directly because it wasn't changed in this PR, but the "redundant closure" warning from `clippy` is caused by #472.